### DISCUSSION TITLE
FlutterFlow - zero-based index pagination

### DISCUSSION THREAD

**Discussion by marcaureln** [pocketbase/pocketbase#3497] [Q&A] [answered]
Hey,

I'm using PocketBase (I'll abbreviate PB) as backend for a mobile app build with FlutterFlow (I'll abbreviate FF, and for those who don't know, it's a kind of no code app builder for Flutter). I'm trying to implement infinite scroll, and FF isn't giving me munch room for maneuver. The indexing start with 0, whereas PB pagination starts with 1. The result in my app is that the first page (or the first items) are duplicate. PB handle the `page=0` as `page=1` (which is great behavior). After, doing some research, I come across `pb_hooks`.

I tried the following :

```js
/// <reference path="../data/types.d.ts" />

onRecordsListRequest((e) => {
    const isFlutterFlow = e.httpContext.queryParam("ff") // console.log => true
    const page = e.httpContext.queryParam("page") // console.log => 0

    if (isFlutterFlow) {
        e.httpContext.setPathParams({ page: page + 1 ?? 1 })
    }
    
}, "myCollection")
```

With this request `api/collections/myCollection/records?sort=-created&page=0&perPage=30&skipTotal=false&ff=true`, I got this:
```
{
  "code": 400,
  "message": "Something went wrong while processing your request.",
  "data": {}
}
```

How can I intercept the request and increment the page query parameter?

**Comment by ganigeorgiev** [comment, marked as answer]
Can't you use [Code Expression](https://docs.flutterflow.io/advanced-functionality/code-expression)?
From the [FlutterFlow docs](https://docs.flutterflow.io/widgets-and-components/widgets/layout-elements/listview#1.-add-paginated-api-call) it seems that it should be possible to define an expression for your API call and combine it with the predefined pagination variables that it provides, something like `var + 1`.

Overwriting the request query parameters unfortunately is not so easy at the moment. `onRecordsListRequest` is triggered after the records retrieval. If you want to overwrite the query parameters you'll have to define a middleware that will either rewrite the request or update directly the parsed `url.Values` from the context.

```js
/// <reference path="../pb_data/types.d.ts" />

// the below middleware will be triggered for every route that has "ff" and "page" query parameters
routerUse((next) => {
    return (c) => {
        const params = c.queryParams();

        if (params.has("ff") && params.has("page")) {
            const page = params.get("page") << 0 // convert to number
            c.queryParams().set("page", page + 1)
        }

        return next(c)
    }
})

```

**Reply by marcaureln** [reply]
Thank you @ganigeorgiev, 

Code Expression actually solve my issue. I have a little overflow, with the `var + 1` we are fetching one more page, but PB is behaving well, simply returning an empty page. I'll stick with that for now. Furthermore, I'm also going to test your middleware, and decide what to keep.

Should I update the discussion title? Seems, that the issue isn't that much about PocketBase.

**Reply by ganigeorgiev** [reply]
> ... simply returning an empty page

This is because of the recent pagination optimizations from https://github.com/pocketbase/pocketbase/discussions/2965.
If you don't need the total* fields that comes by default with the list queries, you can also speed up the query execution by passing `skipTotal=1`.

_I've updated the title of the discussion to mention FlutterFlow in case someone is searching for it._

**Reply by marcaureln** [reply]
Great, I add the `skipTotal` query param, hope that it'll make the infinite scroll experience smoother.

Thanks for everything @ganigeorgiev and have a nice one :v:

**Comment by iamdineshkumar** [comment]
@marcaureln Whether you used pocketbase for authentication also in FF?

**Reply by marcaureln** [reply]
Yes, I also used FF for authentication. You don't have the same experience as Firebase or Supabase (for those who don't know, FF has direct integration for Firebase and Supabase).  Still, you can rely on FF API calls, and they are doing the job. Nothing complicated.

**Reply by iamdineshkumar** [reply]
@marcaureln So Authentication is done using FF API calls and using global params to store session?

**Reply by marcaureln** [reply]
Yeah, you can store the token in `App State`
